File: presence.py
import time
import pickledb
import discord
import os
import requests
import datetime
import logging
from dateutil.tz import tzutc, tzlocal
import re
import asyncio


# runs at 5:00pm PST on sundays, grabs events from https://api.presence.io/wsu/v1/events
# filters by events that are listed for the next 7 days and posts them to the configured channel
def fetch_week_events():
    # log
    logging.info("Fetching events for the next 7 days...")

    raw_events = requests.get("https://api.presence.io/wsu/v1/events").json()

    # filter events to only include events that are listed for the next 7 days
    # use the startDateTimeUtc field to determine it, this field is a datetime string
    # example: 2021-10-13T17:00:00Z
    # the last character is a Z, which means it's in UTC time

    # get the current time in UTC
    current_time = datetime.datetime.utcnow()

    # create a list to store the filtered events
    filtered_events = []

    # loop through all the events
    for event in raw_events:
        # if the event doesn't have a startDateTimeUtc field, skip it
        if "startDateTimeUtc" not in event:
            continue

        # convert the startDateTimeUtc field to a datetime object
        start_time = datetime.datetime.strptime(
            event["startDateTimeUtc"], "%Y-%m-%dT%H:%M:%SZ"
        )

        # calculate the difference between the current time and the event start time
        time_difference = start_time - current_time

        # if the event is within the next 7 days, add it to the list
        if time_difference.days <= 7 and time_difference.days >= 0:
            filtered_events.append(event)

    # sort the events by start time
    filtered_events.sort(
        key=lambda event: datetime.datetime.strptime(
            event["startDateTimeUtc"], "%Y-%m-%dT%H:%M:%SZ"
        )
    )

    # log
    logging.info(f"Found {len(filtered_events)} events for the next 7 days")

    return filtered_events


def post_events(bot, guild_id, server_channels):
    # get the configured channel id for the server
    channel_id = server_channels.get(str(guild_id))

    # if the channel id is not set, send a message to the channel
    if channel_id is None:
        logging.error(
            "No channel has been configured for this server, use /set-channel to set one!"
        )
        return

    # get the channel object from the channel id
    channel = bot.get_channel(channel_id)

    if channel is None:
        logging.error("The channel id is invalid!")
        return

    events = fetch_week_events()

    # post a message to the channel letting the user
    # know what week the bot is posting events for.
    # posting n events for the week of m/d/y

    asyncio.create_task(
        channel.send(
            f"Posting {len(events)} events occurring this next week, after {datetime.datetime.now().strftime('%m/%d/%Y')}"
        )
    )

    # loop through all the events
    for event in events:
        formatted_description = re.sub("<[^<]+?>", "", event["description"])

        if len(formatted_description) > 1024:
            formatted_description = formatted_description[:1021] + "..."

        # build a card embed
        embed = discord.Embed(
            title=event["eventName"],
            # strip HTML from the description (note, this doesn't sanitize)
            description=formatted_description,
            url=f"https://wsu.presence.io/event/{event['uri']}",
            color=0xA60F2D,
        )

        # add the event start time and convert it to a friendly
        # format using the dateutil library
        formatted_time = (
            datetime.datetime.strptime(event["startDateTimeUtc"], "%Y-%m-%dT%H:%M:%SZ")
            .replace(tzinfo=datetime.timezone.utc)
            .astimezone(tz=datetime.timezone(datetime.timedelta(hours=-7)))
            .strftime("%A, %B %d, %Y at %I:%M %p")
        )

        logging.info(event["eventName"] + " | " + formatted_time)
        embed.add_field(
            name="Start Time",
            value=formatted_time,
            inline=False,
        )

        # check if the event has a photo
        if "photoUri" in event:
            # add the event image
            embed.set_image(
                url=f"https://wsu-cdn.presence.io/event-photos/618a42cf-1860-4f89-9fd4-0f42bdbf546c/{event['photoUri']}"
            )

        # check if the event has a location
        if "location" in event:
            # add the location
            embed.add_field(name="Location", value=event["location"], inline=False)

        # check if the event has an organization
        if "organizationName" in event:
            # add the organization
            embed.add_field(
                name="Organization", value=event["organizationName"], inline=False
            )

        # check if the event has an rsvp link
        if "rsvpLink" in event:
            # add the RSVP button
            embed.add_field(
                name="RSVP",
                value=f"[Click Here]({event['rsvpLink']})",
                inline=False,
            )

        # send the embed to the channel
        asyncio.create_task(channel.send(embed=embed))

        # wait 1 second before sending the next embed
        time.sleep(1)

File: presence.ts
import { ActivityType, type Client, type TextChannel } from "discord.js";
import { type PresenceEvent, type ParsedEvent as ParsedPresenceEvent } from "./types";
const { logger } = require("./logger");
import { db } from "./bot";
const cron = require("node-cron");

// Keywords to filter out events
const FILTERED_KEYWORDS = [
    "greek", "fraternity", "sorority", "religious", "church", "temple", "mosque",
    "synagogue", "prayer", "worship", "faith", "god", "jesus", "allah", "buddha",
    "hindu", "christian", "muslim", "jewish", "buddhist", "hinduism", "christianity", "catholic",
    "islam", "judaism", "buddhism", "religious", "spiritual", "spirituality",
    "spiritualism", "spiritualist", "spiritualistic", "spiritualistic", "spiritualistic",
    "kappa", "alpha", "delta", "sigma", "phi", "theta", "omega", "beta", "gamma",
    "zeta", "iota", "lambda", "mu", "nu", "xi", "omicron", "pi", "rho", "tau", "upsilon",
];

/**
 * Get the next week's events from the Presence API
 * @returns A promise that resolves to an array of parsed events
 */
export async function getNextWeekEvents(): Promise<ParsedPresenceEvent[]> {
    // Fetch events from the API
    const response = await fetch("https://api.presence.io/wsu/v1/events");
    // if the response is not OK, throw an error
    if (!response.ok) {
        logger.error(`Failed to fetch events from the Presence API: ${response.statusText}`);
        // increment the presence errors counter
        db.data.stats.presence_errors++;
        db.write();
    }

    const responseJson = await response.json();
    const events: PresenceEvent[] = responseJson;
    
    // first, parse events into a more usable format
    const parsedEvents = events.map((event: PresenceEvent): ParsedPresenceEvent => {
        // dates are of format "%Y-%m-%dT%H:%M:%SZ"
        const startDate = new Date(event.startDateTimeUtc);
        const endDate = new Date(event.endDateTimeUtc);
        if (event.description === null) {
            logger.warn(`Event ${event.eventName} has no description.`);
        }
        return {
            eventName: event.eventName || "No Name",
            location: event.location || "No Location",
            organizationName: event.organizationName || "",
            startDateTimeUtc: startDate || new Date(),
            endDateTimeUtc: endDate || new Date(),
            description: event.description || "No Description",
            isVirtualEventLink: event.isVirtualEventLink || false,
            uri: `https://wsu.presence.io/event/${event.uri}` || "https://wsu.presence.io",
        };
    }).filter((event: ParsedPresenceEvent) => {
        // filter out events that have already ended and are not within the next week
        const now = new Date();
        const nextWeek = new Date();
        nextWeek.setDate(nextWeek.getDate() + 7);
        let res = event.startDateTimeUtc >= now && event.startDateTimeUtc <= nextWeek;
        
        // also filter out greek life events and religious events
        // simply do a search for a list of keywords in the event name, description and organization name
        // if there are any matches, filter out the event
        for (const keyword of FILTERED_KEYWORDS) {
            if (
                event.eventName.toLowerCase().includes(keyword) ||
                event.description.toLowerCase().includes(keyword) ||
                event.organizationName.toLowerCase().includes(keyword)) {
                    res = false;
                    break;
            }
        }

        return res;
    });
    
    return Promise.resolve(parsedEvents);
}

/**
 * Post events for the next week in the configured channels
 * @param client The Discord client
 */
export async function postEvents(client: Client): Promise<void> {
    logger.info("Posting events for the next week...");
    // get the next week's events
    let events = await getNextWeekEvents();
    // update the stats
    db.data.stats.events_this_week = events.length;
    db.write();

    client.user?.setActivity({
        type: ActivityType.Watching,
        name: `for ${events.length} events this week`,
    });

    // sort events by start date
    events.sort((a, b) => a.startDateTimeUtc.getTime() - b.startDateTimeUtc.getTime());
    // limit to 25 events
    events = events.slice(0, 25);

    // log
    logger.info(`Found ${events.length} events for the next week.`);
  
    // get the channel to post events in
    db.data.guild_configs.forEach((config) => {
        const channel = client.channels.cache.get(config.channel_id) as TextChannel;
        if (!channel) {
            logger.warn(`Channel with ID ${config.channel_id} not found. Skipping posting events.`);
            return Promise.reject(`Channel with ID ${config.channel_id} not found.`);
        }
        
        // post the events
        // this will be one big message with all the events as links
        // use discord.js embeds to make it look nice, each event will be a field
        channel.send({
            content: "",
            embeds: [
                {
                    title: events.length > 0 ? `Upcoming ${events.length} Events in the Next Week (Max 25 Shown)` : "No Upcoming Events in the Next Week",
                    // WSU crimson
                    color: 10038562,
                    fields: events.map((event) => {
                        const localStartTime = event.startDateTimeUtc.toLocaleTimeString([], { timeStyle: 'short', timeZone: 'America/Los_Angeles'});
                        const localEndTime = event.endDateTimeUtc.toLocaleTimeString([], { timeStyle: 'short', timeZone: 'America/Los_Angeles'});
                        return {
                            name: `${event.startDateTimeUtc.toDateString()} - ${localStartTime} to ${localEndTime}`,
                            value: `[${event.eventName} - ${event.organizationName}](${event.uri})`,
                        };
                    }),
                    // add "See more" link to the bottom
                    footer: {
                        text: "For more events, visit https://wsu.presence.io",
                    },
                    timestamp: new Date().toISOString(),
                },
            ],
        });
    });

    return Promise.resolve();
}

/**
 * Schedule a job for a guild
 * @param client  The Discord client
 * @param guildId The guild ID to schedule the job for
 */
export async function scheduleJob(client: Client, guildId: string): Promise<void> {

    // schedule a job to post events every Sunday at 10:00 AM
    cron.schedule('0 10 * * 0', () => postEvents(client).catch((e) => logger.error(e)), {
        name: guildId,
        timezone: "America/Los_Angeles"
    });

    logger.info(`Scheduled job for guild ${guildId}`);
}

/**
 * Unschedule a job for a guild
 * @param guildId The guild ID to unschedule the job for
 * @returns void
 */
export async function unscheduleJob(guildId: string): Promise<void> {
    // get tasks
    const tasks = cron.getTasks();

    // find the task with the given name
    for (let [key, value] of tasks.entries()) {
        if (key === guildId) {
            value.stop();
            logger.info(`Unscheduled job for guild ${guildId}`);
            return;
        }
    }
}